File: packages/sdk-python/agios/http.py
"""AGI-OS Python SDK — HTTP Client with retry"""

import time
import json
from typing import Any, Optional
from urllib.request import Request, urlopen
from urllib.error import HTTPError, URLError


class AGIOSError(Exception):
    def __init__(self, message: str, status_code: int, error_type: str = "api_error"):
        super().__init__(message)
        self.status_code = status_code
        self.error_type = error_type


class HTTPClient:
    def __init__(self, base_url: str, api_key: Optional[str] = None,
                 timeout: int = 30, retries: int = 3):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self.retries = retries

    def request(self, method: str, path: str, body: Optional[dict] = None) -> Any:
        url = f"{self.base_url}{path}"
        headers = {"Content-Type": "application/json"}

        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"

        data = json.dumps(body).encode() if body else None
        last_error = None

        for attempt in range(self.retries + 1):
            try:
                req = Request(url, data=data, headers=headers, method=method)
                with urlopen(req, timeout=self.timeout) as resp:
                    response_data = json.loads(resp.read().decode())
                    return response_data.get("data", response_data)

            except HTTPError as e:
                error_body = {}
                try:
                    error_body = json.loads(e.read().decode())
                except Exception:
                    pass

                if e.code == 429 or e.code >= 500:
                    if attempt < self.retries:
                        delay = min(1.0 * (2 ** attempt), 10.0)
                        time.sleep(delay)
                        continue

                error_msg = error_body.get("error", {}).get("message", f"HTTP {e.code}")
                raise AGIOSError(error_msg, e.code)

            except URLError as e:
                last_error = e
                if attempt < self.retries:
                    delay = min(1.0 * (2 ** attempt), 10.0)
                    time.sleep(delay)
                    continue

        raise AGIOSError(f"Request failed: {last_error}", 0)

    def get(self, path: str) -> Any:
        return self.request("GET", path)

    def post(self, path: str, body: Optional[dict] = None) -> Any:
        return self.request("POST", path, body)

File: packages/sdk/src/http.ts
// ═══════════════════════════════════════════════════════
// HTTP Client — Fetch-based with retry & timeout
// ═══════════════════════════════════════════════════════

import { APIError } from './types';

export interface HTTPClientConfig {
  baseUrl: string;
  apiKey?: string;
  timeout: number;
  retries: number;
}

interface RequestOptions {
  method: string;
  path: string;
  body?: unknown;
  headers?: Record<string, string>;
}

export class HTTPClient {
  private config: HTTPClientConfig;

  constructor(config: HTTPClientConfig) {
    this.config = config;
  }

  async request<T>(options: RequestOptions): Promise<T> {
    const url = `${this.config.baseUrl}${options.path}`;
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      ...options.headers,
    };

    if (this.config.apiKey) {
      headers['Authorization'] = `Bearer ${this.config.apiKey}`;
    }

    let lastError: Error | undefined;

    for (let attempt = 0; attempt <= this.config.retries; attempt++) {
      try {
        const controller = new AbortController();
        const timeout = setTimeout(() => controller.abort(), this.config.timeout);

        const response = await fetch(url, {
          method: options.method,
          headers,
          body: options.body ? JSON.stringify(options.body) : undefined,
          signal: controller.signal,
        });

        clearTimeout(timeout);

        if (!response.ok) {
          const errorData = await response.json().catch(() => ({}));
          const apiError: APIError = {
            message: errorData.error?.message || `HTTP ${response.status}`,
            type: errorData.error?.type || 'api_error',
            param: errorData.error?.param,
            code: errorData.error?.code,
          };

          if (response.status === 429 || response.status >= 500) {
            if (attempt < this.config.retries) {
              const delay = Math.min(1000 * Math.pow(2, attempt), 10000);
              await new Promise(r => setTimeout(r, delay));
              continue;
            }
          }

          throw new AGIOSError(apiError.message, response.status, apiError);
        }

        const data = await response.json();
        return (data.data !== undefined ? data.data : data) as T;
      } catch (err) {
        if (err instanceof AGIOSError) throw err;
        lastError = err instanceof Error ? err : new Error(String(err));

        if (attempt < this.config.retries) {
          const delay = Math.min(1000 * Math.pow(2, attempt), 10000);
          await new Promise(r => setTimeout(r, delay));
          continue;
        }
      }
    }

    throw lastError || new Error('Request failed');
  }

  get<T>(path: string): Promise<T> {
    return this.request<T>({ method: 'GET', path });
  }

  post<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>({ method: 'POST', path, body });
  }
}

export class AGIOSError extends Error {
  public statusCode: number;
  public apiError: APIError;

  constructor(message: string, statusCode: number, apiError: APIError) {
    super(message);
    this.name = 'AGIOSError';
    this.statusCode = statusCode;
    this.apiError = apiError;
  }
}
